cpp/mg_utility/mage.hpp

```diff
@@ -0,0 +1,2757 @@
+#pragma once
+
+#include <cstring>
+#include <functional>
+#include <map>
+#include <set>
+#include <string>
+#include <vector>
+
+#include <mg_utils.hpp>
+
+#include "mg_procedure.h"
+#include "mgp.hpp"
+
+namespace mage {
+
+class IndexException : public std::exception {
+ public:
+  explicit IndexException(const std::string &message) : message_(message) {}
+  const char *what() const noexcept override { return message_.c_str(); }
+
+ private:
+  std::string message_;
+};
+
+class ValueException : public std::exception {
+ public:
+  explicit ValueException(const std::string &message) : message_(message) {}
+  const char *what() const noexcept override { return message_.c_str(); }
+
+ private:
+  std::string message_;
+};
+
+class NotEnoughMemoryException : public std::exception {
+ public:
+  const char *what() const throw() { return "Not enough memory!"; }
+};
+
+namespace util {
+// uint to int conversion in C++ is a bit tricky. Take a look here
+// https://stackoverflow.com/questions/14623266/why-cant-i-reinterpret-cast-uint-to-int
+// for more details.
+template <typename TDest, typename TSrc>
+TDest MemcpyCast(TSrc src) {
+  TDest dest;
+  static_assert(sizeof(dest) == sizeof(src), "MemcpyCast expects source and destination to be of the same size");
+  static_assert(std::is_arithmetic<TSrc>::value, "MemcpyCast expects source to be an arithmetic type");
+  static_assert(std::is_arithmetic<TDest>::value, "MemcpyCast expects destination to be an arithmetic type");
+  std::memcpy(&dest, &src, sizeof(src));
+  return dest;
+}
+}  // namespace util
+
+// Forward declarations
+class Nodes;
+using GraphNodes = Nodes;
+class GraphRelationships;
+class Relationships;
```

Comment:
What's the difference between `GraphRelationships` and `Relationships`?

Reply:
In the existing C API one cannot iterate over all relationships in the graph, only over those adjacent to a node (see `vertex_iter_in_edges` and `vertex_iter_out_edges`).
`GraphRelationships` implements iterating over all relationships in the graph, whereas `Relationship` is just a wrapper over `mgp_edges_iterator`.